crates/matrix-sdk-base/src/deserialized_responses.rs

```diff
@@ -67,6 +70,170 @@ pub struct AmbiguityChanges {
+        match (&self.decancered, &other.decancered) {
+            (None, None) => self.raw == other.raw,
+            (None, Some(_)) | (Some(_), None) => false,
+            (Some(this), Some(other)) => this == other,
+        }
+    }
+}
+
+impl DisplayName {
+    /// Regex pattern matching an MXID.
+    const MXID_PATTERN: &str = "@.+:.+";
+
+    /// Regex pattern matching some left-to-right formatting marks:
+    ///     * LTR and RTL marks U+200E and U+200F
+    ///     * LTR/RTL and other directional formatting marks U+202A - U+202F
+    const LEFT_TO_RIGHT_PATTERN: &str = "[\u{202a}-\u{202f}\u{200e}\u{200f}]";
+
+    /// Regex pattern matching bunch of unicode control characters and otherwise
+    /// misleading/invisible characters.
+    ///
+    /// This includes:
+    ///     * various width spaces U+2000 - U+200D
+    ///     * Combining characters U+0300 - U+036F
+    ///     * Blank/invisible characters (U2800, U2062-U2063)
+    ///     * Arabic Letter RTL mark U+061C
+    ///     * Zero width no-break space (BOM) U+FEFF
+    const HIDDEN_CHARACTERS_PATTERN: &str =
+        "[\u{2000}-\u{200D}\u{300}-\u{036f}\u{2062}-\u{2063}\u{2800}\u{061c}\u{feff}]";
+
+    /// Creates a new [`DisplayName`] from the given raw string.
+    ///
+    /// The raw display name is transformed into a Unicode-normalized form, with
+    /// common confusable characters removed to reduce ambiguity.
+    ///
+    /// **Note**: If removing confusable characters fails,
+    /// [`DisplayName::is_ambiguous`] will return `true`, and
+    /// [`DisplayName::as_str()`] will return the original raw string provided
+    /// during construction.
+    pub fn new(raw: &str) -> Self {
+        let normalized = raw.nfd().collect::<String>();
+        let replaced = HIDDEN_CHARACTERS_REGEX.replace_all(&normalized, "");
+
+        let decancered = decancer::cure!(&replaced)
+            .ok()
+            .map(|cured| LEFT_TO_RIGHT_REGEX.replace_all(cured.as_ref(), "").to_string());
+
+        Self { raw: raw.to_owned(), decancered }
+    }
+
+    /// Is this display name considered to be ambiguous?
+    ///
+    /// A display name is classified as ambiguous if it resembles an MXID or
+    /// if it uses a combination of left-to-right characters and [homoglyphs](https://en.wikipedia.org/wiki/Homoglyph)
+    /// to mimic other similar display names, even if only one user has this
+    /// name.
+    pub fn is_ambiguous(&self) -> bool {
+        // If we have cancer or look like an MXID, then we're ambiguous.
+        self.has_cancer() || self.might_look_like_an_mxid() || self.has_left_to_right_characters()
+    }
+
+    /// Returns the underlying normalized and sanitized string of this
+    /// [`DisplayName`].
+    ///
+    /// If normalization and removal of confusable characters failed, this will
+    /// return the original string provided during the construction of the
+    /// [`DisplayName`].
+    pub fn as_str(&self) -> &str {
```

Comment:
The doc comment is lying to me at first ("returns the underlying normalised..." when there is actually an exception in which it doesn't return the normalised string) and the name of the method is far too benign for its magical behaviour. Can we do away with `as_str` and use `as_normalised_str` instead?

Reply:
I removed the method: https://github.com/matrix-org/matrix-rust-sdk/pull/4237/commits/6f08ae43020acf25c3d6f0da612f38a588be3314